File: backend/app/api/v1/endpoints/site_settings.py
from app.models import (
    SiteSetting,
    KB_DOCUMENT_DEFAULT_MAX_UPLOAD_SIZE_MB,
    KB_DOCUMENT_MIN_MAX_UPLOAD_SIZE_MB,
    KB_DOCUMENT_MAX_MAX_UPLOAD_SIZE_MB,
)
from app.schemas.site_setting import PublicSiteSettingsResponse
from app.schemas.response import Response, success
from fastapi import APIRouter

router = APIRouter()

THEME_MODE_VALUES = {"system", "light", "dark"}
THEME_BRANDING_DISPLAY_VALUES = {"full", "name_only", "icon_only", "hidden"}


def _normalize_enum(value: object, allowed_values: set[str], default: str) -> str:
    if isinstance(value, str) and value in allowed_values:
        return value
    return default


def _normalize_hex_color(value: object) -> str:
    if not isinstance(value, str):
        return ""
    stripped = value.strip()
    if len(stripped) not in {4, 7} or not stripped.startswith("#"):
        return ""
    hex_digits = stripped[1:]
    if all(char in "0123456789abcdefABCDEF" for char in hex_digits):
        return stripped
    return ""


def _normalize_kb_document_max_upload_size_mb(value: object) -> int:
    if not isinstance(value, int) or isinstance(value, bool):
        return KB_DOCUMENT_DEFAULT_MAX_UPLOAD_SIZE_MB
    if value < KB_DOCUMENT_MIN_MAX_UPLOAD_SIZE_MB:
        return KB_DOCUMENT_MIN_MAX_UPLOAD_SIZE_MB
    if value > KB_DOCUMENT_MAX_MAX_UPLOAD_SIZE_MB:
        return KB_DOCUMENT_MAX_MAX_UPLOAD_SIZE_MB
    return value


@router.get("/public", response_model=Response[PublicSiteSettingsResponse])
async def get_public_settings():
    """Get public site settings (no authentication required)"""
    settings = await SiteSetting.get_all_by_category(public_only=True)
    return success(
        data=PublicSiteSettingsResponse(
            site_name=settings.get("site_name", "Clouisle"),
            site_description=settings.get("site_description", ""),
            site_url=settings.get("site_url", ""),
            site_icon=settings.get("site_icon", ""),
            auth_page_layout=settings.get("auth_page_layout", "centered"),
            theme_mode=_normalize_enum(
                settings.get("theme_mode"), THEME_MODE_VALUES, "system"
            ),
            theme_primary_color=_normalize_hex_color(
                settings.get("theme_primary_color")
            ),
            theme_primary_foreground_color=_normalize_hex_color(
                settings.get("theme_primary_foreground_color")
            ),
            theme_branding_display=_normalize_enum(
                settings.get("theme_branding_display"),
                THEME_BRANDING_DISPLAY_VALUES,
                "full",
            ),
            icp_record_number=settings.get("icp_record_number") or "",
            icp_record_url=settings.get("icp_record_url") or "",
            terms_enabled=bool(settings.get("terms_enabled")),
            terms_url=settings.get("terms_url") or "",
            terms_text=settings.get("terms_text") or "",
            privacy_enabled=bool(settings.get("privacy_enabled")),
            privacy_url=settings.get("privacy_url") or "",
            privacy_text=settings.get("privacy_text") or "",
            require_terms_acceptance_on_register=bool(
                settings.get("require_terms_acceptance_on_register")
            ),
            allow_registration=settings.get("allow_registration", True),
            require_approval=settings.get("require_approval", False),
            email_verification=settings.get("email_verification", True),
            enable_captcha=settings.get("enable_captcha", False),
            allow_account_deletion=settings.get("allow_account_deletion", True),
            sso_enabled=settings.get("sso_enabled", False),
            sso_allow_password_login=settings.get("sso_allow_password_login", True),
            kb_document_max_upload_size_mb=_normalize_kb_document_max_upload_size_mb(
                settings.get(
                    "kb_document_max_upload_size_mb",
                    KB_DOCUMENT_DEFAULT_MAX_UPLOAD_SIZE_MB,
                )
            ),
        )
    )

File: frontend/lib/api/site-settings.ts
import { api } from './client'

export type AuthPageLayout = 'centered' | 'split'
export type ThemeMode = 'system' | 'light' | 'dark'
export type ThemeBrandingDisplay = 'full' | 'name_only' | 'icon_only' | 'hidden'

export interface SiteSetting {
  key: string
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  value: any
  value_type: string
  category: string
  description?: string
  is_public: boolean
}

export interface SiteSettings {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  settings: Record<string, any>
}

export interface PublicSiteSettings {
  site_name: string
  site_description: string
  site_url: string
  site_icon: string
  auth_page_layout: AuthPageLayout
  theme_mode: ThemeMode
  theme_primary_color: string
  theme_primary_foreground_color: string
  theme_branding_display: ThemeBrandingDisplay
  icp_record_number: string
  icp_record_url: string
  terms_enabled: boolean
  terms_url: string
  terms_text: string
  privacy_enabled: boolean
  privacy_url: string
  privacy_text: string
  require_terms_acceptance_on_register: boolean
  allow_registration: boolean
  require_approval: boolean
  email_verification: boolean
  enable_captcha: boolean
  allow_account_deletion: boolean
  sso_enabled: boolean
  sso_allow_password_login: boolean
  kb_document_max_upload_size_mb: number
}

export interface GeneralSettings {
  site_name: string
  site_description: string
  site_url: string
  site_icon: string
  default_language: string
  auth_page_layout: AuthPageLayout
  theme_mode: ThemeMode
  theme_primary_color: string
  theme_primary_foreground_color: string
  theme_branding_display: ThemeBrandingDisplay
  icp_record_number: string
  icp_record_url: string
  terms_enabled: boolean
  terms_url: string
  terms_text: string
  privacy_enabled: boolean
  privacy_url: string
  privacy_text: string
  require_terms_acceptance_on_register: boolean
}

export interface SecuritySettings {
  allow_registration: boolean
  require_approval: boolean
  email_verification: boolean
  allow_account_deletion: boolean
  default_role_id: string
  default_team_id: string
  default_team_role: 'viewer' | 'member' | 'admin'
  min_password_length: number
  require_uppercase: boolean
  require_number: boolean
  require_special_char: boolean
  session_timeout_days: number
  single_session: boolean
  max_login_attempts: number
  lockout_duration_minutes: number
  enable_captcha: boolean
  sso_enabled: boolean
  sso_allow_password_login: boolean
  sso_auto_create_users: boolean
  sso_require_approval: boolean
  sso_match_by_email: boolean
  // Password expiration
  password_expiration_enabled: boolean
  password_expiration_days: number
  password_expiration_warning_days: number
  password_history_count: number
  password_min_age_days: number
  force_password_change_first_login: boolean
  // TOTP
  require_totp: boolean
}

export interface EmailSettings {
  smtp_enabled: boolean
  smtp_host: string
  smtp_port: number
  smtp_encryption: 'none' | 'ssl' | 'tls'
  smtp_username: string
  smtp_password: string
  email_from_name: string
  email_from_address: string
}

export interface DingTalkSettings {
  dingtalk_enabled: boolean
  dingtalk_notification_type: 'webhook' | 'app'
  dingtalk_webhook_url: string
  dingtalk_secret: string
  dingtalk_app_key: string
  dingtalk_app_secret: string
  dingtalk_agent_id: string
}

export interface WeChatSettings {
  wechat_enabled: boolean
  wechat_notification_type: 'webhook' | 'app'
  wechat_webhook_url: string
  wechat_corp_id: string
  wechat_agent_id: string
  wechat_secret: string
}

export interface FeishuSettings {
  feishu_enabled: boolean
  feishu_notification_type: 'webhook' | 'app'
  feishu_webhook_url: string
  feishu_secret: string
  feishu_app_id: string
  feishu_app_secret: string
}

export interface WebhookSettings {
  webhook_enabled: boolean
  webhook_url: string
  webhook_method: 'POST' | 'GET'
  webhook_headers: Record<string, string>
  webhook_body_template: string
  webhook_secret: string
}

export interface SlackSettings {
  slack_enabled: boolean
  slack_webhook_url: string
}

export interface AutoNotificationConfig {
  channels: string[]  // Global channels for all enabled notifications
  enabled_types: string[]  // Enabled notification types
}

export const siteSettingsApi = {
  /**
   * Get public site settings (no auth required)
   */
  async getPublic(): Promise<PublicSiteSettings> {
    return api.get<PublicSiteSettings>('/site-settings/public')
  },
}
